### PR TITLE
There will be cloudbuilders (inside freight)

### PR DESCRIPTION
**Pull request by egsy**
Adds a Check to find a corresponding sha build in Google Cloud Container Builder.

### CODE REVIEW

**Comment by egsy** [comment]
please leave more comments when you have a moment @JTCunning~

File: freight/checks/cloudbuilder.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            sha {str} -- required, commit SHA of build to check
+            config {dict} -- optional dict to pass additional args
+        """
+        api_root = "https://cloudbuild.googleapis.com/v1/projects/{}/builds".format(
+            config["project"]
+        )
+
+        oauth_command = "gcloud auth application-default print-access-token"
+        try:
+            oauth_token = config["oauth_token"]
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Don't wrap this is a try/except, gracefully `get` the object and then check against it being null.

File: freight/checks/cloudbuilder.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            "filter": "sourceProvenance.resolvedRepoSource.commitSha={}".format(sha)
+        }
+        headers = {
+            "Accepts": "application/json",
+            "Authorization": "Bearer {}".format(oauth_token)
+        }
+
+        resp = http.get(api_root, headers=headers, params=params)
+        if resp.status_code != 200:
+            print("[ ERROR {} ]\tNo data for build present".format(resp.status_code))
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Remove this print statement.

File: freight/checks/cloudbuilder.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            raise CheckFailed("No data for build present")
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Add the response message into this exception message you're raising here.

File: freight/checks/cloudbuilder.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        build_data = resp.json()
+        build_id = build_data["builds"][0]["id"]
+        build_status = build_data["builds"][0]["status"]
+        build_url = build_data["builds"][0]["logUrl"]
+        build_logs = build_data["builds"][0]["logsBucket"].lstrip("gs:/")
+        gcloudstatus = {
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Add another statement below the `SUCCESS` one that just raises a `CheckFailed` and returns the status and the message associated with it.

File: freight/checks/cloudbuilder.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            "FAILURE": "Build or step failed to complete successfully.",
+            "INTERNAL_ERROR": "Build or step failed due to an internal cause.",
+            "TIMEOUT": "Build or step took longer than was allowed.",
+            "CANCELLED": "Build or step was canceled by a user."
+        }
+
+        if build_status == "FAILURE":
+            build_logtext = "https://storage.googleapis.com/{}/log-{}.txt".format(build_logs, build_id)
+            log = http.get(build_logtext, headers=headers)
+            print("[ {} ]\t{}\nSee details: {}".format(build_status, gcloudstatus["FAILURE"], build_url))
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Remove these two print statements and add the content to the exception message instead.

File: freight/checks/cloudbuilder.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            print("Printing log...\n\n\n{}".format(log.text))
+            raise CheckFailed()
+
+        if build_status == "WORKING":
+            print("[ {} ]\t{}\nSee details: {}".format(build_status, gcloudstatus["WORKING"], build_url))
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Remove this print statement and add the content to the exception message instead.

File: freight/checks/cloudbuilder.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            raise CheckPending("build incomplete")
+
+        if build_status == "SUCCESS":
+            print("[ {} ]\t{}\nSee details: {}".format(build_status, gcloudstatus["SUCCESS"], build_url))
```
Comment (JTCunning) [inline freight/checks/cloudbuilder.py]: Remove this print statement.

File: tests/checks/test_cloudbuilder.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        self.app = self.create_app(repository=self.repo)
+        self.test_project = "mycoolproject"
+        self.test_sha = "0987654321"
+        self.test_token = "mysuperfaketoken"
+
+
+class CloudbuilderContextCheckTest(CloudbuilderCheckBase):
+    @responses.activate
+    def test_build_success(self):
+        id = "successful_build_id"
```
Comment (JTCunning) [inline tests/checks/test_cloudbuilder.py]: `id` is a built-in. Choose a different variable name.

File: tests/checks/test_cloudbuilder.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        body = json.dumps({
+            "builds": [{
+                "id": "{}".format(id),
```
Comment (JTCunning) [inline tests/checks/test_cloudbuilder.py]: You're formatting a blank string with a string. You can shorten this to `"id": build_id`.

File: tests/checks/test_cloudbuilder.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                "logUrl": "https://console.cloud.google.com/gcr/builds/{}?project={}".format(id, self.test_project),
+                "logsBucket": "gs://{}.cloudbuild-logs.googleusercontent.com".format(self.test_project),
+                "status": "SUCCESS",
+            },
+            ]
+        })
+        responses.add(responses.GET, "https://cloudbuild.googleapis.com/v1/projects/{}/builds".format(self.test_project), body=body)
+
+        config = {
+            "contexts": ["cloudbuilder"],
```
Comment (JTCunning) [inline tests/checks/test_cloudbuilder.py]: Search all of your code and remove the notion of `context(s)`. This is GitHub specific.

File: tests/checks/test_cloudbuilder.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            "https://cloudbuild.googleapis.com/v1/projects/{}/builds".format(self.test_project),
+            body=body
+        )
+
+        config = {
+            "contexts": ["cloudbuilder"],
+            "project": self.test_project,
+            "oauth_token": self.test_token}
+
+        with pytest.raises(CheckFailed):
```
Comment (JTCunning) [inline tests/checks/test_cloudbuilder.py]: This `with` statement should be moved so that the only method you're calling inside of it is the check.

File: freight/checks/cloudbuilder.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        resp = http.get(api_root, headers=headers, params=params)
+        if resp.status_code != 200:
+            print("[ ERROR {} ]\tNo data for build present".format(resp.status_code))
+            raise CheckFailed("No data for build present")
+
+        build_data = resp.json()
+        build_id = build_data["builds"][0]["id"]
+        build_status = build_data["builds"][0]["status"]
+        build_url = build_data["builds"][0]["logUrl"]
+        build_logs = build_data["builds"][0]["logsBucket"].lstrip("gs:/")
```
Comment (mattrobenolt) [inline freight/checks/cloudbuilder.py]: `lstrip` here is probably going to do something you don't want it to do. `lstrip` is actually working on the string as an iterable, and it strips individual characters, not the sequence.

So like:

```python
>>> 'gs://sssssssggggg-io'.lstrip('gs:/')
'-io'
>>> 'gs://sssssssggggg-io'.lstrip('s/:g')
'-io'
```

**Comment by egsy** [comment]
@JTCunning have made the requested changes! please review again~

File: freight/checks/cloudbuilder.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+            "WORKING": "Build or step is being executed.",
+            "SUCCESS": "Build or step finished successfully.",
+            "FAILURE": "Build or step failed to complete successfully.",
+            "INTERNAL_ERROR": "Build or step failed due to an internal cause.",
+            "TIMEOUT": "Build or step took longer than was allowed.",
+            "CANCELLED": "Build or step was canceled by a user.",
+        }
+
+        if build_status == "FAILURE":
+            print(build_id)
```
Comment (mattrobenolt) [inline freight/checks/cloudbuilder.py]: I assume you should remove this `print()`

File: freight/checks/cloudbuilder.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+            build_logtext = "https://storage.googleapis.com/{}/log-{}.txt".format(
+                build_logs[slice(5, None)], build_id
+            )
+            log = http.get(build_logtext, headers=headers)
+            raise CheckFailed(
+                "[ {} ]\t{}\nSee details: {}\nPrinting log...\n\n\n{}".format(
```
Comment (mattrobenolt) [inline freight/checks/cloudbuilder.py]: I assume these will break with unicode. Any of these where you do `.format()` make sure the string you're using is a unicode.

So:

```python
raise CheckFailed(
  u"...".format(...)
)
```

instead. Otherwise, you will very very likely cause an exception.

File: tests/checks/test_cloudbuilder.py [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+                self.test_project
+            ),
+            body=body,
+        )
+
+        config = {"project": self.test_project, "oauth_token": self.test_token}
+
+        with pytest.raises(CheckFailed) as exception_info:
+            self.check.check(self.app, self.test_sha, config)
+        print (exception_info.value)
```
Comment (mattrobenolt) [inline tests/checks/test_cloudbuilder.py]: Remove these print()s as well from tests.

File: freight/checks/cloudbuilder.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+            "FAILURE": "Build or step failed to complete successfully.",
+            "INTERNAL_ERROR": "Build or step failed due to an internal cause.",
+            "TIMEOUT": "Build or step took longer than was allowed.",
+            "CANCELLED": "Build or step was canceled by a user.",
+        }
+
+        if build_status == "FAILURE":
+            print(build_id)
+            build_logtext = "https://storage.googleapis.com/{}/log-{}.txt".format(
+                build_logs[slice(5, None)], build_id
```
Comment (mattrobenolt) [inline freight/checks/cloudbuilder.py]: nit, just use `[5:]` this is odd as `slice()`. I think you got this from me when I mentioned this, but I would not suggest using this unless it was required. And here, it's not.

File: freight/checks/cloudbuilder.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        api_root = "https://cloudbuild.googleapis.com/v1/projects/{}/builds".format(
+            config["project"]
+        )
+
+        oauth_command = "gcloud auth application-default print-access-token"
+
+        oauth_token = config.get("oauth_token")
+        if oauth_token is None:
+            oauth_token = subprocess.check_output(shlex.split(oauth_command)).rstrip()
```
Comment (mattrobenolt) [inline freight/checks/cloudbuilder.py]: Why are we using `shlex` to split this apart instead of just doing:

```python
subprocess.check_output(['gcloud', 'auth', 'application-default', 'print-access-token'])
```
